src/diskusage.cpp

```diff
@@ -0,0 +1,245 @@
+/*
+ * Copyright (C) 2015 Jolla Ltd.
+ * Contact: Thomas Perl <thomas.perl@jolla.com>
+ *
+ * You may use this file under the terms of the BSD license as follows:
+ *
+ * "Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are
+ * met:
+ *   * Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *   * Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in
+ *     the documentation and/or other materials provided with the
+ *     distribution.
+ *   * Neither the name of Nemo Mobile nor the names of its contributors
+ *     may be used to endorse or promote products derived from this
+ *     software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE."
+ */
+
+#include "diskusage.h"
+#include "diskusage_p.h"
+
+#include <QThread>
+#include <QDir>
+#include <QProcess>
+#include <QDebug>
+#include <QJSEngine>
+
+#include <sys/statvfs.h>
+
+DiskUsageWorker::DiskUsageWorker(QObject *parent)
+    : QObject(parent)
+    , m_quit(false)
+{
+}
+
+DiskUsageWorker::~DiskUsageWorker()
+{
+}
+
+static quint64 calculateSize(QString directory)
+{
+    // In lieu of wordexp(3) support in Qt, fake it
+    if (directory.startsWith("~/")) {
+        directory = QDir::homePath() + '/' + directory.mid(2);
+    }
+
+    QDir d(directory);
+    if (!d.exists() || !d.isReadable()) {
+        return 0L;
+    }
+
+    // TODO: Recursively sum up disk usage manually while avoiding symlinks
+    // instead of exec'ing du for every app
+
+    QProcess du;
+    du.start("du", QStringList() << "-sxb" << directory, QIODevice::ReadOnly);
+    du.waitForFinished();
+    if (du.exitStatus() != QProcess::NormalExit) {
```

Comment:
QProcess status reporting is non-consistent: e.g. if process was not started because executable does not exists exitStatus will be NormalExit. It seems the only reliable way is to connect to QProcess signals. Also error() is set to UnknownError by default.